### PR TITLE
[2.7] [Form] fix choice value "false" in ChoiceType

### PR DESCRIPTION
**Pull request by HeahDude**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17292, #14712, #17789 |
| License | MIT |
| Doc PR | - |
- [x] Add tests for choices with `boolean` and `null` values, and with a placeholder
- [x] Fix FQCN in 2.8 tests, see #17759
- [x] Remove `choices_as_values` in 3.0 tests, see #17886

### CODE REVIEW

**Comment by HeahDude** [comment]
Tests failure is not related to this PR.

**Comment by xabbuh** [comment]
:+1: 

Status: Reviewed

ping @symfony/deciders

**Comment by fabpot** [comment]
:+1:

**Comment by webdevilopers** [comment]
Fix works! :+1:

**Comment by Tobion** [comment]
This fix contradicts https://github.com/symfony/symfony/pull/16681/files#r45997531
But I think @webmozart was forgetting that `false` would then conflict with the placeholder optional value.
So changing `false` to `0` looks better to me indeed.
But then `castableToString` must also be updated to reflect that change.

Please add a test for that.

```
'choices' => [
        'Yes'  => true,
        'No'   => false,
        'Undecided'   => null,
    ],
```

null and false should not conflict and thus can both be used as `''` and `'0'` value (instead of fallback solution to use incrementing numbers).

**Comment by Tobion** [comment]
Btw, this would fix `false` choices with `required => false`. But a `null` choice with required false, would still have the same problem?! IMO a placeholder value should always be different than any given choice.

**Comment by Tobion** [comment]
So basically we need to map all scalars that would result in an empty string to something else to not conflict we the placeholder:

I would propose this mapping:

- `false` -> `0`
- `null` -> `~`
- `''` -> `-`

And if there are still conflicts e.g. because someone has both `''` and `'-'` and choices (which would conflict with above mapping), then it still uses the incrementing numbers as fallback which is what `castableToString` ensures.

**Comment by HeahDude** [comment]
Thanks @Tobion for the review. You're right, `castableToString` must be refactored so when using both `boolean` and `null` values it returns true for the test you mention pass.

I'm fixing it.

Status: Needs Work

**Comment by HeahDude** [comment]
Tests are fixed.

Status: Needs Review

File: src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/ChoiceToValueTransformerTest.php [hunk 1]
```diff
@@ -34,8 +34,8 @@ public function transformProvider()
     {
         return array(
             // more extensive test set can be found in FormUtilTest
-            array('', '0'),
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/ChoiceToValueTransformerTest.php]: Should it be `array(null, '0')`?

File: src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/ChoiceToValueTransformerTest.php [hunk 2]
```diff
@@ -34,8 +34,8 @@ public function transformProvider()
...
-            array(false, '1'),
+            array('', ''),
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/ChoiceToValueTransformerTest.php]: Should it be `array(null, '')` ?

File: src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/ChoiceToValueTransformerTest.php [hunk 3]
```diff
@@ -52,9 +52,9 @@ public function reverseTransformProvider()
         return array(
             // values are expected to be valid choice keys already and stay
             // the same
-            array('0', ''),
-            array('1', false),
-            array('2', 'X'),
+            array('', ''),
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/ChoiceToValueTransformerTest.php]: Same here.

File: src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/ChoicesToValuesTransformerTest.php [hunk 4]
```diff
@@ -32,7 +32,7 @@ protected function tearDown()
     public function testTransform()
     {
         $in = array('', false, 'X');
-        $out = array('0', '1', '2');
+        $out = array('', '0', 'X');
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/ChoicesToValuesTransformerTest.php]: And here `array(null, false, 'X')` ?

File: src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/ChoicesToValuesTransformerTest.php [hunk 5]
```diff
@@ -53,7 +53,7 @@ public function testTransformExpectsArray()
     public function testReverseTransform()
     {
         // values are expected to be valid choices and stay the same
-        $in = array('0', '1', '2');
+        $in = array('', '0', 'X');
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/ChoicesToValuesTransformerTest.php]: Would it work here with `null` instead of an empty string ?

**Comment by Tobion** [comment]
The problem now is

```
'choices' => [
    'Undecided'   => null,
    'Yes'  => true,
    'No'   => false,
],
'placeholder' => 'Please choose',
```

will not add the placeholder anymore because `null` is mapped to `''` value and thus the placeholder is ignored because the choice assumes the null is already the placeholder value. And even worse is that people cannot select `Undecided` in this required select anymore because the browser will treat it as the placeholder as well (cuz it has an empty value). Before it used incrementing integers and thus was not affected by this.
That is why I suggested to change the mapping from `null` to something like `~` as value to not be treated as placeholder.

**Comment by HeahDude** [comment]
@Tobion, thanks again for the heads-up, so I keep on testing.

Status: Needs Work

File: src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php [hunk 6]
```diff
@@ -233,13 +233,13 @@ private function castableToString(array $choices, array &$cache = array())
                 }
 
                 continue;
-            } elseif (!is_scalar($choice)) {
+            } elseif (!is_scalar($choice) && null !== $choice) {
```
Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php]: @Tobion a solution for the placeholder issue is to revert the change on this line so it keeps the old behaviour when `null` is a value option. WDUT ?

Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php]: I confirm this would work.

But @Tobion is still right, when having both a placeholder and a choice with an explicit `null` or empty string `''` value, there is a conflict.

The default `null` choice is pre selected over the `placeholder` added in the choice list, so perhaps it should not be added in this case special case :
if default data is `null` is && `''` in choice values then `placeholder` would be set to `false`.
There would be no need then to reverse `elseif (!is_scalar($choice) && null !== $choice)` in a new commit.
To have a choice pre selected __and__ a placeholder looks odd to me, especially with select input.
Also the use case for a `null` value in a choice list make me guess it's about handling a default value in some way.
Does it makes sense to you ? ping __@symfony/deciders__

Otherwise, @Tobion suggested this mapping :
>     false -> '0'
    null -> '~'

This would solve it indeed, but may not result in an expected behaviour when using custom submission (e.g javascript) as one would send an empty value and want it to be mapped to the choice with a `null` value instead of needing to send `'~'`.

The only other way would I can see would be to set a special string value for the `placeholder` in order to prevent it from being submitted and to fallback on default data . Just for the example something like "~sf_placeholder-value" that we know will not conflict with other string values.

Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php]: Actually, I really think there is a good reason to not reverse this change.

This `placeholder` issue is not related to this PR, in 2.7+, if one of the choice has a value `null`, and `null` is the default data, `placeholder` is never selected by default in favour of this choice.

Plus, if you submit the form with the `placeholder` selected and even if `required` is `true` the form is valid and fallback on the `null` data when synchronising which may be a kind of bug as it would not be expected.

The PR fixes it as `null` is always converted to empty string so the `placeholder` is not added in the choice list as `placeholder_in_choices` is `true`.

To avoid the conflict, setting `placeholder` with an internal default string value like `~placeholder~` might be considered after all...

Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php]: This issue #14607 seems related.

**Comment by HeahDude** [comment]
Status: Needs Review

File: src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php (hunk 6 above)
Comment (Tobion) [inline src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php]: > This would solve it indeed, but may not result in an expected behaviour when using custom request submission (e.g javascript, api) as one would send an empty value and want it to be mapped to the choice with a null value instead of needing to send '~'.

The counter argument seems wrong. If people used `null`` as a choice, then it used numeric keys for values. So javascript, apis etc would also need to send the numeric keys (depending on the order) and not an empty string.

Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php]: @Tobion thanks, got it, so there is no need to use `~` for `null` value, if I revert the condition above as it will always use numeric values, is it ?

Comment (Tobion) [inline src/Symfony/Component/Form/ChoiceList/ArrayChoiceList.php]: yes lets revert the null handling.

**Comment by HeahDude** [comment]
@Tobion thanks for confirming :)

Status: Needs Work

**Comment by Tobion** [comment]
I think https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/DataTransformer/ChoiceToValueTransformer.php#L42 should only accept null and string.

Accepting scalar seems unecessary and also conflicts with `false` -> `'0'` behavior as the string cast below is disregarding this.

**Comment by HeahDude** [comment]
@Tobion `ChoiceToValueTransformer` is a `ViewTransformer` and is called after the model data of the `ArrayChoiceList` is set, meaning at this time `false` is already casted to `'0'`.

So I don't if it's worth changing it.

I've reverted `null` being castable to string in commit 5dabbd5.

All my tests handling a `false` value pre selected are passing except the two with an expanded single choice list.

I've reverted the "hack" I used to get this one passing (see https://github.com/symfony/symfony/commit/5dabbd5406025ee8091b07ec93059a847b94de54#diff-11ee91a32c601e8e1d509117556d53b3L150), because it was not working if `null` and `false` are both in choices as they use in this case a numerical index value (so `false` may be !== `'0'`) but it would fail with any object anyway as it depends on `castableToString`.

I've finally managed to get it fix but it implies a little refactoring of `RadioListMapper::mapDataToForms()`.
AFAIK it is only used by the `ChoiceList` in this special case of expanded and not multiple, I think we should do it but it might be a BC break as this method "fixed" will expect a "value" instead of a "choice" as argument.

I don't know if there is some use cases where some one overrides or extends this class or even use it independently but it seems to be the only way to fix those last two tests see e3b366e.

Status: Needs Review

**Comment by Tobion** [comment]
you are right that `ChoiceToValueTransformer` is not called with boolean. That is exactly the reason why it should also not be allowed to. Otherwise it allows things that are not consistent and behave wrong if misused.

**Comment by HeahDude** [comment]
@Tobion it may be not be called for `null` either at this point, shouldn't it accept strings only ?

**Comment by HeahDude** [comment]
@Tobion also you said :

>Btw, this would fix false choices with required => false. But a null choice with required false, would still have the same problem?! IMO a placeholder value should always be different than any given choice.

I don't fully understand what you mean by that, is there anything more I can try to fix or a test to add ?

Thanks again for you feedback so far, much appreciated.

**Comment by Tobion** [comment]
>  it may be not be called for null either at this point, shouldn't it accept strings only ?

Normally empty strings are converted to null by the form component. So I would assume ChoiceToValueTransformer is called with null. So I guess we have to keep the null behavior.

> I don't fully understand what you mean by that, is there anything more I can try to fix or a test to add ?

That was when I thought null is mapped to `''` (same as false). But it uses numeric indexes then as we found out. So you can ignore that.

**Comment by HeahDude** [comment]
Rebased and squashed, all comments addressed.

I left the `RadioListMapper` refactoring in a different commit because it solves #17789 independently in it may not be the best way to do it so I can discard it and we can keep tracking this bug in #17889.
Be aware that merging only the first commit will add some failing tests for this.

I also left the `ChoiceType` tests addition in a separated commit to ease the rebase of #17759 and  #17886 updating them for 2.8 and 3.0.

Needs a final review.

**Comment by HeahDude** [comment]
Test failure fix by #17859

**Comment by fabpot** [comment]
Thank you @HeahDude.

**Comment by Taluu** [comment]
This seems to break API ways on Symfony 2.8.3. When before I could do the following :

```php
<?php
class ChoiceType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder->add('status', SfChoiceType::class, [
            'choices'  => [true, false, 'true', 'false'],
            'choices_as_values' => true
        ]);
    }
}
```

With the following json sent data :

```json
{
        "choices":[
          {
            "hash":"foo4u",
            "label":"foo",
            "status": true
          },

          {
            "hash":"bar4u",
            "label":"bar",
            "status": "false"
          },

          {
            "hash":"baz4u",
            "label":"baz",
            "status": false
          }
        ]
}
```

whilre before I could use the values `true`, `"true"`, `false` and `"false"` I can now use only `true` and `false`.

**Comment by Taluu** [comment]
And when I change the order in the `choices` key in the options with `['true', 'false', true, false]` instead, all my `false` are changed to `null`...

**Comment by HeahDude** [comment]
@Taluu, The submission of a `ChoiceType` actually requires the string value to get properly mapped to the model choice data.

If you submit a true value it will be cast to string so `"1"` not by the choice list constructor like in this fix but directly in the `Form::submit()` method. Changing the behavior at this point is a BC break.

So `false` get casted to empty string which cannot mapped the `false` model choice data expecting the string '0' as set in the choice list constructor (by this PR).

If the is `null` or an object among the choices, the default `ArrayChoiceList::castableToString()` method will return false then you should rely on numeric index to send as string or int in a json hash.

`false` value could get in conflict with placeholder in some cases.

Another way to solve this is to return false in `castableToString()` is there is boolean but we loose the ability to match `true` as '"1"' with is a bigger BC break...

You can rely on `choice_value` option and always send the good strings values on submission, or use `choice_label` option and a choices array with a numeric index and submit the right index.

**Comment by Taluu** [comment]
For legacy reason we have these `"true"` and `"false"` values. But we have an internal ticket to use real booleans instead ("one day my prince will come"), and then using a `CheckboxType` would be the way to go.

Until then, we will have to stick with `"true"` and `"false"`. Using either `['true', 'false']` or `[true, false]` works though...

**Comment by HeahDude** [comment]
Edge case then. Happy coding :)

**Comment by Taluu** [comment]
Yep, I confirm, this is indeed due to this. Don't know if it was there or not before, but using a `CheckboxType` when we will have fixed the internal ticket (And chosen between string or boolean) will fix the damn thing. Yeay.

```
Symfony\Component\Validator\ConstraintViolation
Object(Symfony\Component\Form\Form).children[choices].children[0].children[status] = 1
Caused by:
Symfony\Component\Form\Exception\TransformationFailedException
Unable to reverse value for property path "[status]": The choice "1" does not exist or is not unique
Caused by:
Symfony\Component\Form\Exception\TransformationFailedException
The choice "1" does not exist or is not unique
```

And using directly the `CheckboxType` is not possible (Even though it passes with flying colors - a `"false"` is transformed into `false` and `"true"` into `true`), as we have a need to have `"true"` and `"false"` values.  So yeah nothing to report.

**Comment by HeahDude** [comment]
@Taluu, note that the `ChoiceType` when expanded uses `CheckboxType` or `RadioType` (which extends the first) to represent as a virtual boolean if the actual string value is selected or not as a choice among other values and to be mapped back to a model value.

That's why it messes with PATCH method and `$clearMissing` (ref : https://github.com/symfony/symfony/pull/17771#issuecomment-188156241).

When you need to use `CheckboxType` or `RadioType` to map a real boolean property in a form, consider using a `CollectionType`.

**Comment by webmozart** [comment]
@Taluu You could preprocess your request data and postprocess your responses to transform Booleans to strings and back. That would at least fix your API.
